### PR TITLE
Always stop hc worker actor after sending results.

### PR DESCRIPTION
**Pull request by ConnorDoyle**
Fixes an actor leak in the health check subsystem.

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/health/HealthCheckWorkerActor.scala [hunk 1]
```diff
@@ -31,9 +31,7 @@ class HealthCheckWorkerActor extends Actor with ActorLogging {
     case HealthCheckJob(task, check) =>
       val replyTo = sender() // avoids closing over the volatile sender ref
 
-      val replyWithHealth = doCheck(task, check)
-
-      replyWithHealth.onComplete {
+      doCheck(task, check).andThen {
```
Comment (drexin) [inline src/main/scala/mesosphere/marathon/health/HealthCheckWorkerActor.scala]: This is a little inconsistent. Here you use a `.` and in the second `andThen` you don't. Any specific reason for this?

File: src/main/scala/mesosphere/marathon/health/HealthCheckWorkerActor.scala [hunk 2]
```diff
@@ -41,9 +39,7 @@ class HealthCheckWorkerActor extends Actor with ActorLogging {
             task.getVersion,
             s"${t.getClass.getSimpleName}: ${t.getMessage}"
           )
-
-          context stop self
-      }
+      } andThen { case _ => self ! PoisonPill }
```
Comment (drexin) [inline src/main/scala/mesosphere/marathon/health/HealthCheckWorkerActor.scala]: As you don't use the result, I think `onComplete` should be used here.

**Comment by ConnorDoyle** [comment]
@drexin addressed review comments.

**Comment by drexin** [comment]
Can we add a test for this?

**Comment by ConnorDoyle** [comment]
Added a test to verify worker actor shutdown, improved the existing test for this actor a bit in the process.

File: src/test/scala/mesosphere/marathon/health/HealthCheckWorkerActorTest.scala [hunk 3]
```diff
@@ -13,33 +13,70 @@ import scala.concurrent.duration._
...
-    }
-    finally {
-      socket.close()
-    }
+    ref ! HealthCheckJob(task, HealthCheck(protocol = Protocol.TCP))
+
+    try { Await.result(res, 1.seconds) }
+    finally { socket.close() }
+
+    val probe = TestProbe()
```
Comment (drexin) [inline src/test/scala/mesosphere/marathon/health/HealthCheckWorkerActorTest.scala]: This shouldn't be necessary, `TestKit` provides the same functionality.

**Comment by drexin** [comment]
Oh, the build failed with an error I haven't seen before. This is probably not related to your changes. Could you please verify and open a ticket?

**Comment by drexin** [comment]
Besides my comment this looks good. Feel free to merge afterwards.

**Comment by ConnorDoyle** [comment]
Filed #946 for the flaky test.

File: src/test/scala/mesosphere/marathon/health/HealthCheckWorkerActorTest.scala (hunk 3 above)
Comment (ConnorDoyle) [inline src/test/scala/mesosphere/marathon/health/HealthCheckWorkerActorTest.scala]: Thanks, removed the probe and the import.

**Comment by drexin** [comment]
Thanks!

**Comment by ConnorDoyle** [comment]
Thanks for the reviews, merging now.

**Pull request merged**